src/core/trulens/core/utils/deprecation.py

```diff
@@ -0,0 +1,164 @@
+"""Utilities for handling deprecation."""
+
+from enum import Enum
+import functools
+import inspect
+from typing import Any, Callable, Dict, Iterable, Optional, Type, Union
+import warnings
+
+from trulens.core.utils import imports as imports_utils
+
+PACKAGES_MIGRATION_LINK = (
+    "https://trulens.org/docs/migration-guide"  # TODO: update link
+)
+
+
+def packages_dep_warn(module: str):
+    """Issue a deprecation warning for a backwards-compatibility modules.
+
+    This is specifically for the trulens_eval -> trulens module renaming and
+    reorganization.
+    """
+
+    warnings.warn(
+        f"The `{module}` module is deprecated. "
+        f"See {PACKAGES_MIGRATION_LINK} for instructions on migrating to `trulens.*` modules.",
+        DeprecationWarning,
+        stacklevel=3,
+    )
+
+
+def has_deprecated(obj: Union[Callable, Type]) -> bool:
+    """Check if a function or class has been deprecated."""
+
+    return has_moved(obj)
+
+
+def has_moved(obj: Union[Callable, Type]) -> bool:
+    """Check if a function or class has been moved."""
+
+    return (
+        hasattr(obj, "__doc__")
+        and obj.__doc__ is not None
+        and "has moved:\n" in obj.__doc__
+    )
+
+
+def function_moved(func: Callable, old: str, new: str):
+    """Issue a warning upon function call that has been moved to a new location.
+
+    Issues the warning only once.
+    """
+
+    assert hasattr(
```

Comment:
Is using assert a good way in deprecation code?

Reply:
unsure, have to give some error, change to raise TypeError to make it more clear that this is a static-type of error instead of a runtime type.